### ISSUE TITLE
Unorthodox  Sorting / Coloring

### ISSUE THREAD

**Issue by eriksonssilva** [dfinke/ImportExcel#1076]
Hello Doug! I hope you're doing fine!
So, I've been using your module quite a lot now, and I dare to say it saved me HUNDREDS of hours of work!
I have a question and I think the Import-Excel is the answer, but I could not find it in any document, nor reading the help...
I have this scenario: I receive a spreadsheet to input into the system.
The problem: It has HUNDREDS of entries and they are not in order (Not Sorted.xlsx), like this:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/V-SILV~2/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/V-SILV~2/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Title | Dep Code | Group
-- | -- | --
Title 1 | Dep Code 7 | Group 9
Title 1 | Dep Code 5 | Group 9
Title 1 | Dep Code 8 | Group 9
Title 2 | Dep Code 9 | Group 9
Title 2 | Dep Code 9 | Group 3
Title 2 | Dep Code 2 | Group 3
Title 2 | Dep Code 9 | Group 4
Title 3 | Dep Code 11 | Group 7
Title 3 | Dep Code 1 | Group 7
Title 3 | Dep Code 9 | Group 7

</body>

</html>

And seeing like this, might be easy to sort by hand. But this is only one example... I receive files with hundreds, even thousands of lines. And what I need to do is to:
1 - Sort column A from A to Z
2 - Filter column A with Title 1 and column B with Dep Code 1. Then I copy the outcome to a new spreadsheet.
Repeat the process with Title 1 + Dep Code 2, Title 2 and color it with a new color. Title 1 + Dep 3, new color.
So on and so forth, for hours and hours.

[Not Sorted.xlsx](https://github.com/dfinke/ImportExcel/files/7145179/Not.Sorted.xlsx)

So I'd like to see if its possible to have it sorted like this:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/V-SILV~2/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/V-SILV~2/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Title | Dep Code | Group
-- | -- | --
Title 1 | Dep Code 3 | Group 3
Title 1 | Dep Code 3 | Group 7
Title 1 | Dep Code 4 | Group 7
Title 2 | Dep Code 9 | Group 3
Title 2 | Dep Code 9 | Group 4
Title 2 | Dep Code 9 | Group 11
Title 2 | Dep Code 9 | Group 7
Title 3 | Dep Code 1 | Group 7
Title 3 | Dep Code 9 | Group 9
Title 3 | Dep Code 9 | Group 3

</body>

</html>
Basically if Title 1 + Dep Code 1, the color will be Blank, if Title 1 + Dep Code 2, will be yellow.
You can refer to the "Sorted Colored.xlsx" to see what I mean by color sort.
No need to worry with the "group" column (although would be good).

[Sorted Colored.xlsx](https://github.com/dfinke/ImportExcel/files/7145180/Sorted.Colored.xlsx)

Do you think that would be possible?

Thanks a lot in advance!

Cheers,

Erikson

**Comment by dfinke** [comment]
Thank you @eriksonssilva . `Import-Excel` can read that data into PowerShell so you can sort and manipulate it. Off the top of my head, you can then export it back to excel and use `Set-ExcelRange` to color the rows. You would need to implement that logic to figure where the "breaks" in the data are and then apply the color.

**Comment by eriksonssilva** [comment]
Thanks a lot, Doug! I really appreciate you getting back to me.
Yeah, now that you gave the hint away, I feel dumb haha
I have resolved my issue this way:
I'll export my spreadsheet using Import-Excel (Of course).
Once I have it, I'll then add each row to an array
So "row 1" = $array[0],  "row 2" = $array[1] yadda yadda yadda
Even if the row has 10 columns, it'll occupy only one position of the array... Why? 
Because that way I'll sort my text the exact way I wish!

I am using a regex to sort to "natural" readable sorting. So "1, 2, 3... 10, 11", instead of "1, 10, 11, 2, 3".
The regex is this: 
`$toNatural= { [regex]::Replace($_, '\d+',{$args[0].Value.Padleft(20)})}`

That done, I'll use my first column to get all the repeated values. If the next value is the same, color will also be the same.
If it is different, then the color will be changed.

Well, might sound a little bit gibberish, bit I am only starting programming, so I may sound senseless.
Here is my code so you understand better.

`$toNatural= { [regex]::Replace($_, '\d+',{$args[0].Value.Padleft(20)})}


$get = Import-excel .\Not.Sorted.xlsx

$firstColumn = $get.firstColumn
$secondColumn = $get.secondColumn
$thirdColumn = $get.thirdColumn
$array = @()


foreach ($i in $firstColumn.Length) {

    $array += $firstColumn[$i] + " | " + $secondColumn[$i] + " | " + $thirdColumn[$i]

}


$sorted = $array | Sort-Object $toNatural

$r = Get-Random -Maximum 254
$g = Get-Random -Maximum 254
$b = Get-Random -Maximum 254

$color = New-ConditionalText  -BackgroundColor ([System.Drawing.Color]::FromArgb($r, $g, $b)) -ForeGroundColor Black

$lastdata = $firstColumn[0]

rm .\Book1.xlsx

for($i = 1; $i -lt $firstColumn.Length + 1; $i++) {




      if($firstColumn[$i] -eq $lastdata){
      
          #$firstColumn[$i] | Export-Excel Book1.xlsx -StartRow $i -StartColumn 1 -ConditionalText  $color
          #$secondColumn[$i] | Export-Excel Book1.xlsx -StartRow $i -StartColumn 2 -ConditionalText $color
          #$thirdColumn[$i] | Export-Excel Book1.xlsx -StartRow $i -StartColumn 3 -ConditionalText $color
          $sorted[$i]| Sort-Object $toNatural | Export-Excel Book1.xlsx -StartRow $i -StartColumn 4 -ConditionalText $color
            
      }
      
      else {
            
          $r = Get-Random -Maximum 254
          $g = Get-Random -Maximum 254
          $b = Get-Random -Maximum 254

          $color = New-ConditionalText  -BackgroundColor ([System.Drawing.Color]::FromArgb($r, $g, $b)) -ForeGroundColor Black
      
          #$firstColumn[$i]| Export-Excel Book1.xlsx -StartRow $i -StartColumn 1 -ConditionalText $color
          #$secondColumn[$i] | Export-Excel Book1.xlsx -StartRow $i -StartColumn 2 -ConditionalText $color
          #$thirdColumn[$i] | Export-Excel Book1.xlsx -StartRow $i -StartColumn 3 -ConditionalText $color
          $sorted[$i]| Sort-Object $toNatural |Export-Excel Book1.xlsx -StartRow $i -StartColumn 4 -ConditionalText $color

      
      }    

          $lastdata = $firstColumn[$i]

}`

I have used a random RGB color, because this way it's almost 100% sure I'll never have the same color.
And that's it. Thanks a lot for pointing me towards the right direction!
I'm closing this issue, if you don't mind.

Cheers!
